crates/cargo-wdk/src/actions/new/mod.rs

```diff
@@ -291,3 +291,552 @@ impl<'a> NewAction<'a> {
+        set_expectations_fn: impl FnOnce(TestSetup) -> TestSetup,
+        assert_fn: impl FnOnce(Result<(), NewActionError>),
+    ) {
+        let test_setup = TestSetup::new(path);
+        let test_setup = set_expectations_fn(test_setup);
+
+        let result = NewAction::new(
+            path,
+            driver_type,
+            verbosity_level,
+            &test_setup.mock_exec,
+            &test_setup.mock_fs,
+        )
+        .run();
+
+        assert_fn(result);
+    }
+
+    /// Enum representing different steps where failures can be injected during
+    /// tests. Used to configure mock expectations for specific failure
+    /// scenarios.
+    enum FailureStep {
+        CargoNew(Output),
+        CopyLibRsTemplate,
+        UpdateCargoToml(bool, bool, bool),
+        CreateInxFile,
+        CopyBuildRsTemplate,
+        CopyCargoConfig,
+    }
+
+    /// Test helper struct that provides a fluent API for configuring mock
+    /// providers.
+    ///
+    /// This struct holds mock implementations of `CommandExec` and `Fs`
+    /// providers along with the driver project path. It provides methods to
+    /// set up mock expectations for various stages of the driver project
+    /// creation workflow.
+    ///
+    /// # Usage
+    ///
+    /// Create a new instance with [`TestSetup::new`], then chain calls to
+    /// configure mocks for each step of the workflow. Use
+    /// [`TestSetup::set_expectations_with`] to configure all mocks at once
+    /// based on an optional failure point, or call individual `expect_*`
+    /// methods for fine-grained control.
+    ///
+    /// # Example
+    ///
+    /// ```ignore
+    /// let test_setup = TestSetup::new(Path::new("test_driver"))
+    ///     .set_expectations_with(Some(FailureStep::CopyLibRsTemplate), None);
+    ///
+    /// // Now use test_setup.mock_exec and test_setup.mock_fs in your test
+    /// ```
+    struct TestSetup<'a> {
+        /// The path to the driver project being created in the test.
+        path: &'a Path,
+        /// Mock implementation of the command execution provider.
+        mock_exec: MockCommandExec,
+        /// Mock implementation of the file system provider.
+        mock_fs: MockFs,
+    }
+
+    impl<'a> TestSetup<'a> {
+        fn new(path: &'a Path) -> Self {
+            Self {
+                path,
+                mock_exec: MockCommandExec::new(),
+                mock_fs: MockFs::new(),
+            }
+        }
+
+        fn set_expectations_with(
+            mut self,
+            failure_step: Option<FailureStep>,
+            expected_flag: Option<String>,
+        ) -> Self {
+            if let Some(FailureStep::CargoNew(override_output)) = failure_step {
+                return self.expect_cargo_new(Some(override_output), expected_flag);
+            }
+            self = self.expect_cargo_new(None, expected_flag);
+
+            if matches!(failure_step, Some(FailureStep::CopyLibRsTemplate)) {
```

Comment:
Nit: use `if let Some(...) = failure_step` for all these checks instead of `matches!`